### PR TITLE
Refactor AVL tree

### PR DESCRIPTION
**Pull request by ljw7975**
Refactor insert/delete methods
clean up code

### CODE REVIEW

File: src/main/java/Tree/AVLTree.java [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        updateHeight((AVLNode<T>) current);
+
+        int heightDifference = getHeightDifference((AVLNode<T>) current);
+        if (heightDifference > 1 || heightDifference < -1) {
+            balanceTreeAfterInsertion(insert_key, (AVLNode<T>) current, heightDifference);
+        }
+    }
+
+    protected void updateHeightAfterDeletion(AVLNode<T> node) {
+        assert node == null : "Height of null node cannot be updated";
```
Comment (ljj7975) [inline src/main/java/Tree/AVLTree.java]: `input node cannot be null`

File: src/main/java/Tree/AVLTree.java [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+                    current = null;
+                } else {
+                    int parent_key = parent_node.getKey();
+                    if (delete_key < parent_key) {
+                        parent_node.setLeftNode(null);
+                        updateHeightAfterDeletion(parent_node);
+                    } else {
+                        parent_node.setRightNode(null);
+                        updateHeightAfterDeletion(parent_node);
+                    }
```
Comment (ljj7975) [inline src/main/java/Tree/AVLTree.java]: bring out `updateHeight` outside of if

File: src/main/java/Tree/AVLTree.java [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    protected void rightRotation() {
+        AVLNode<T> newRoot = (AVLNode<T>) current.getLeftNode();
+
+        if (newRoot.getRightNode() != null) {
+            current.setLeftNode(newRoot.getRightNode());
+        } else {
+            current.setLeftNode(null);
+        }
+
+        if (current.getParentNode() != null) {
```
Comment (ljj7975) [inline src/main/java/Tree/AVLTree.java]: Can you use LinkNodes we defined in BST?

File: src/main/java/Tree/AVLTree.java [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    protected void leftRotation() {
+        AVLNode<T> newRoot = (AVLNode<T>) current.getRightNode();
+
+        if (newRoot.getLeftNode() != null) {
+            current.setRightNode(newRoot.getLeftNode());
+        } else {
+            current.setRightNode(null);
+        }
+
+        if (current.getParentNode() != null) {
```
Comment (ljj7975) [inline src/main/java/Tree/AVLTree.java]: Again, try to reimplement this with linkNodes

File: src/main/java/Tree/BinarySearchTree.java [hunk 5]
```diff
@@ -63,25 +63,32 @@ public T getValue(int key) {
     protected BinaryNode<T> getSuccessor (BinaryNode<T> node) {
         assert node == null : "Input node must not be null";
 
-        node = node.getRightNode();
-        while (node.getLeftNode() != null) {
-            node = node.getLeftNode();
+        if (node.getRightNode() != null) {
```
Comment (ljj7975) [inline src/main/java/Tree/BinarySearchTree.java]: why is this change necessary?

File: src/main/java/Tree/BinarySearchTree.java [hunk 6]
```diff
@@ -63,25 +63,32 @@ public T getValue(int key) {
...
         return node;
     }
 
     protected BinaryNode<T> getPredecessor (BinaryNode<T> node) {
         assert node == null : "Input node must not be null";
 
-        node = node.getLeftNode();
-        while (node.getRightNode() != null) {
-            node = node.getRightNode();
+        if (node.getLeftNode() != null) {
```
Comment (ljj7975) [inline src/main/java/Tree/BinarySearchTree.java]: same here

File: src/main/java/Tree/BinarySearchTree.java (hunk 5 above)
Comment (ljw7975) [reply src/main/java/Tree/BinarySearchTree.java]: when node = null, node.getLeftNode() call produces nullpointexception

File: src/main/java/Tree/BinarySearchTree.java (hunk 6 above)
Comment (ljw7975) [reply src/main/java/Tree/BinarySearchTree.java]: above

File: src/main/java/Tree/AVLTree.java [hunk 7]
```diff
@@ -144,36 +146,37 @@ public void delete(int delete_key) {
...
         if (newRoot.getRightNode() != null) {
-            current.setLeftNode(newRoot.getRightNode());
+            linkNodes(current, newRoot.getRightNode(), 'L');
         } else {
-            current.setLeftNode(null);
+            linkNodes(current, null, 'L');
         }
 
         if (current.getParentNode() != null) {
             if (current == current.getParentNode().getLeftNode()) {
```
Comment (ljj7975) [inline src/main/java/Tree/AVLTree.java]: have a variable for `current.getParentNode()` before this line
parent node is being repeatedly used.

Same for the other rotation
